src/parser/ion_1_0/text/mod.rs

```diff
@@ -0,0 +1,2614 @@
+                |(year, month)| ion::TextTimestamp::new(ion::Date::Month { year, month }, None),
+            ),
+            map(terminated(take_year, one_of("tT")), |year| {
+                ion::TextTimestamp::new(ion::Date::Year { year }, None)
+            }),
+        )),
+        ion::Timestamp::Text,
+    )(i)?;
+
+    Ok((i, timestamp))
+}
+
+/// fragment
+/// DATE
+///     : YEAR '-' MONTH '-' DAY
+///     ;
+fn take_date(i: &str) -> IonResult<&str, (i32, u8, u8)> {
+    let (i, (y, (m, d))) = pair(
+        take_year,
+        pair(
+            preceded(char('-'), take_month),
+            preceded(char('-'), take_day),
+        ),
+    )(i)?;
+
+    Ok((i, (y, m, d)))
+}
+
+/// fragment
+/// YEAR
+///     : '000'                     [1-9]
+///     | '00'            [1-9]     DEC_DIGIT
+///     | '0'   [1-9]     DEC_DIGIT DEC_DIGIT
+///     | [1-9] DEC_DIGIT DEC_DIGIT DEC_DIGIT
+///     ;
+fn take_year(i: &str) -> IonResult<&str, i32> {
+    map(
+        alt((
+            recognize(pair(tag("000"), one_of("123456789"))),
+            recognize(tuple((
+                tag("00"),
+                one_of("123456789"),
+                one_if(is_dec_digit),
+            ))),
+            recognize(tuple((
+                tag("0"),
+                one_of("123456789"),
+                take_while_m_n(2, 2, is_dec_digit),
+            ))),
+            recognize(tuple((
+                one_of("123456789"),
+                take_while_m_n(3, 3, is_dec_digit),
+            ))),
+        )),
+        |s: &str| {
+            s.parse::<i32>()
+                .expect("the parser ensures it will be within range")
+        },
+    )(i)
+}
+
+/// fragment
+/// MONTH
+///     : '0' [1-9]
+///     | '1' [0-2]
+///     ;
+fn take_month(i: &str) -> IonResult<&str, u8> {
+    map(
+        alt((
+            recognize(pair(char('0'), one_of("123456789"))),
+            recognize(pair(char('1'), one_of("012"))),
+        )),
+        |s: &str| {
+            s.parse::<u8>()
+                .expect("the parser ensures it will be within range")
+        },
+    )(i)
+}
+
+/// fragment
+/// DAY
+///     : '0'   [1-9]
+///     | [1-2] DEC_DIGIT
+///     | '3'   [0-1]
+///     ;
+fn take_day(i: &str) -> IonResult<&str, u8> {
+    map(
+        alt((
+            recognize(pair(char('0'), one_of("123456789"))),
+            recognize(pair(one_of("12"), one_if(is_dec_digit))),
+            recognize(pair(char('3'), one_of("01"))),
+        )),
+        |s: &str| {
+            s.parse::<u8>()
+                .expect("the parser ensures it will be within range")
+        },
+    )(i)
+}
+
+fn take_hour_and_minute(i: &str) -> IonResult<&str, (u32, u32)> {
+    separated_pair(take_hour, char(COLON), take_minute)(i)
+}
+
+type ParsedTime = ((u32, u32), Option<(u32, Option<BigInt>)>, UtcOffset);
+
+fn assemble_time(((hour, minute), maybe_second, offset): ParsedTime) -> ion::Time {
+    let (second, maybe_fractional) = if maybe_second.is_none() {
```

Comment:
Perhaps split this into separate functions with/without a `second` parameter

Reply:
Done, and cleaned it up with some matches.